### PR TITLE
Adds SQL support for collections of entities

### PR DESCRIPTION
**Pull request by csilvestrini**

### CODE REVIEW

File: java/arcs/android/common/SqlUtils.kt [hunk 1]
```diff
@@ -61,12 +60,28 @@ inline fun Cursor.forEach(block: (Cursor) -> Unit) {
...
+ * Usage:
+ *
+ * ```kotlin
+ * val name: String? = database.rawQuery("SELECT name FROM users WHERE id = ?", emptyArray(123))
+ *     .forSingleResult { it.getString(0) }
+ * ```
+ */
+inline fun <T> Cursor.forSingleResult(block: (Cursor) -> T?): T? = use {
+    require(count == 0 || count == 1) { "Expected 0 or 1 results, found $count." }
+    if (moveToFirst()) block(it) else null
```
Comment (jasonwyatt) [inline java/arcs/android/common/SqlUtils.kt]: Nicely done!

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 2]
```diff
@@ -142,14 +152,48 @@ class DatabaseImpl(
...
+            rawQuery(
+                """
+                    SELECT entity_refs.entity_id, entity_refs.backing_storage_key
+                    FROM collection_entries
+                    JOIN entity_refs ON collection_entries.value_id = entity_refs.id
+                    WHERE collection_entries.collection_id = ?
+                """.trimIndent(),
+                arrayOf(collectionId.toString())
+            ).forEach {
+                values.add(Reference(
```
Comment (jasonwyatt) [inline java/arcs/android/storage/database/DatabaseImpl.kt]: Nit: place `Reference(` on a new line, and insert a new line between the `))` on line 177.

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 3]
```diff
@@ -184,6 +228,51 @@ class DatabaseImpl(
...
+            val schemaTypeId = getSchemaTypeId(data.schema, db)
+
+            // Retrieve existing collection ID for this storage key, if one exists.
+            var collectionId = getCollectionId(storageKey, db)
+
+            if (collectionId == null) {
+                // Create a new collection ID and storage key ID.
+                collectionId = insert(TABLE_COLLECTIONS, null, ContentValues().apply {
+                    put("type_id", schemaTypeId)
+                })
```
Comment (jasonwyatt) [inline java/arcs/android/storage/database/DatabaseImpl.kt]: Nit: use "place arguments on separate lines"

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 4]
```diff
@@ -184,6 +228,51 @@ class DatabaseImpl(
...
+                insert(TABLE_STORAGE_KEYS, null, ContentValues().apply {
+                    put("storage_key", storageKey.toString())
+                    put("data_type", DataType.Collection.ordinal)
+                    put("value_id", collectionId)
+                })
```
Comment (jasonwyatt) [inline java/arcs/android/storage/database/DatabaseImpl.kt]: Nit: use "place arguments on separate lines"

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 5]
```diff
@@ -184,6 +228,51 @@ class DatabaseImpl(
...
+            val content = ContentValues().apply {
+                put("collection_id", collectionId)
+            }
+            // TODO: Don't do this one-by-one.
+            data.values
+                .map { getEntityReferenceId(it, db) }
+                .forEach { referenceId ->
+                    insert(TABLE_COLLECTION_ENTRIES, null, content.apply {
+                        put("value_id", referenceId)
+                    })
```
Comment (jasonwyatt) [inline java/arcs/android/storage/database/DatabaseImpl.kt]: Nit: use "place arguments on separate lines"

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 6]
```diff
@@ -256,30 +346,58 @@ class DatabaseImpl(
...
+        db.transaction {
+            val refId = rawQuery(
+                "SELECT id FROM entity_refs WHERE entity_id = ? AND backing_storage_key = ?",
+                arrayOf(reference.id, reference.storageKey.toString())
+            ).forSingleResult { it.getLong(0) }
+            refId ?: run {
+                insert(TABLE_ENTITY_REFS, null, ContentValues().apply {
+                    put("entity_id", reference.id)
+                    put("backing_storage_key", reference.storageKey.toString())
+                })
```
Comment (jasonwyatt) [inline java/arcs/android/storage/database/DatabaseImpl.kt]: Nit: use "place arguments on separate lines"

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 7]
```diff
@@ -333,16 +451,11 @@ class DatabaseImpl(
...
-                fieldValueId
-            } else {
-                val content = ContentValues().apply {
+            ).forSingleResult { it.getLong(0) }
+            fieldValueId ?: run {
+                insert(tableName, null, ContentValues().apply {
                     put("value", valueStr)
-                }
-                insert(tableName, null, content)
+                })
```
Comment (jasonwyatt) [inline java/arcs/android/storage/database/DatabaseImpl.kt]: Nit: use "place arguments on separate lines"

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 8]
```diff
@@ -353,10 +466,9 @@ class DatabaseImpl(
         fun runSelectQuery(tableName: String) = db.rawQuery(
             "SELECT value FROM $tableName WHERE id = ?",
             arrayOf(valueId.toString())
-        ).use {
-            require(it.moveToFirst()) { "Unknown primitive with ID $valueId." }
-            it.getString(0)
-        }
+        ).forSingleResult { it.getString(0) }
+            ?: throw IllegalArgumentException("Unknown primitive with ID $valueId.")
```
Comment (jasonwyatt) [inline java/arcs/android/storage/database/DatabaseImpl.kt]: Definitely optional: I wonder if it'd be good to add another extension function called `requireSingleResult`?

File: java/arcs/core/storage/StorageKey.kt [hunk 9]
```diff
@@ -24,4 +24,12 @@ abstract class StorageKey(val protocol: String) {
         childKeyWithComponent("handle/$handleId")
 
     override fun toString(): String = "$protocol://${toKeyString()}"
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageKey.kt]: This won't compile cross-platform..   I suggest just doing `if (other == null || other::class != this::class) return false`

File: java/arcs/android/storage/database/DatabaseImpl.kt (hunk 2 above)
Comment (csilvestrini) [reply java/arcs/android/storage/database/DatabaseImpl.kt]: Done

File: java/arcs/android/storage/database/DatabaseImpl.kt (hunk 3 above)
Comment (csilvestrini) [reply java/arcs/android/storage/database/DatabaseImpl.kt]: Done

File: java/arcs/android/storage/database/DatabaseImpl.kt (hunk 4 above)
Comment (csilvestrini) [reply java/arcs/android/storage/database/DatabaseImpl.kt]: Done

File: java/arcs/android/storage/database/DatabaseImpl.kt (hunk 7 above)
Comment (csilvestrini) [reply java/arcs/android/storage/database/DatabaseImpl.kt]: Done

File: java/arcs/android/storage/database/DatabaseImpl.kt (hunk 8 above)
Comment (csilvestrini) [reply java/arcs/android/storage/database/DatabaseImpl.kt]: I thought about that, but I couldn't figure out a nice looking way of supplying the lambda and the error message

```
.requireSingleResult("Unknown primitive ...") {
    it.getString(0)
}
```

It seems backwards to me, but the alternative means inlining the lambda, which isn't heaps nice either...

```
.requireSingleResult(
    { it.getString(0) },
    "Unknown primitive ..."
)
```

File: java/arcs/android/storage/database/DatabaseImpl.kt (hunk 5 above)
Comment (csilvestrini) [reply java/arcs/android/storage/database/DatabaseImpl.kt]: Done

File: java/arcs/android/storage/database/DatabaseImpl.kt (hunk 6 above)
Comment (csilvestrini) [reply java/arcs/android/storage/database/DatabaseImpl.kt]: Done

File: java/arcs/core/storage/StorageKey.kt (hunk 9 above)
Comment (csilvestrini) [reply java/arcs/core/storage/StorageKey.kt]: Done

**Comment by csilvestrini** [comment]
Presubmit cl/293728692 failing for unrelated reasons

**Pull request merged**